File: Question9/Question9.py
import itertools
memo = None
def recursion(client,mask,banker,t):
    if memo[banker][client][mask] != -1:
        return memo[banker][client][mask]
    if client == len(t[0]):
        return 0
    min_cost = float("Inf")
    #print("\nb -> ",banker,"c->",client,"bin ",bin(mask))
    for j in range(len(t)):#banker loop
        if (mask & (1<<j) == 0 or j==banker):
        #print(client,j,mask,len(memo),len(memo[0]),len(memo[0][0]))
            bit = mask | (1 << j)
            a = recursion(client+1,bit,j,t)
            min_cost = min(min_cost,t[j][client] + a)
        #print(t[j][client],end = " ") 
    memo[banker][client][mask] = min_cost
    return min_cost
            
def assignment(cost):
    min_cost = float('inf')
    for i in range(len(cost)):
        #print("banker ",i)
        time = recursion(0,0,i,cost)
        #print(time)
        min_cost = min(min_cost,time)
    return min_cost

def BankerAndClients(c, b, t):
    global memo
    memo =[[[-1 for i in range(pow(2,b)+1)]for j in range(c+1)]for k in range(b+1)]
    min_time = assignment(t)
    return min_time
	# Participants code will be here


if __name__ == "__main__":
	abc=list(map(int,input().split()))
	client=abc[0]
	banker=abc[1]
	time=[[0 for i in range(20)] for j in range(10)]
	time = [list(map(int, input().split())) for y in range(banker)]
	result = BankerAndClients(client, banker, time)
	# Do not remove below line
	print(result)    
	# Do not print anything after this line# your code goes here

File: Question9/Question9.c
#include <stdio.h>

int memo[16][32][2048] = {{{NULL}}};
int b,c;
int g_time[16][32];
int recursion(int client,int banker,int key){
    if (memo[banker][client][key] != NULL)return memo[banker][client][key];
    if (client == c)return 0;
    int curr_min_t = 2147483647;
    for(int j =0;j<b;j++){
        if ((key & (1<<j))==0 || j==banker){
            int bit = key | (1<<j);
            int a = recursion(client+1,j,bit);
            if((g_time[j][client] + a)<curr_min_t){
                 curr_min_t = g_time[j][client] + a;
            }
        }
    }
    memo[banker][client][key] = curr_min_t;
    return curr_min_t;
}
int main(){
    scanf("%d %d", &c, &b);
    for (int i=0; i<b; i++){
        for(int j = 0;j < c;j++){
          scanf("%d",&g_time[i][j]);
        }
    }
	int min_t = 2147483647;
    for (int i=0;i<b;i++){
        int _time = recursion(0,i,0);
        if(min_t>_time){
            min_t = _time;
        }
    }
    // Do not remove below line
    printf("%d",min_t);
    // Do not print anything after this line
}
